internal/app/upstream/client.go

```diff
@@ -55,11 +74,177 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+// TODO: pass retry configurations
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+// NewMockClient creates a mock implementation for testing
+func NewMockClient(
+	ctx context.Context,
+	ldsClient v2.ListenerDiscoveryServiceClient,
+	rdsClient v2.RouteDiscoveryServiceClient,
+	edsClient v2.EndpointDiscoveryServiceClient,
+	cdsClient v2.ClusterDiscoveryServiceClient,
+	callOptions CallOptions) Client {
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}
+}
+
+func (m *client) OpenStream(ctx context.Context, request *v2.DiscoveryRequest) (<-chan *Response, error) {
+	var stream grpc.ClientStream
+	var err error
+	switch request.GetTypeUrl() {
+	case listenerTypeURL:
+		stream, err = m.ldsClient.StreamListeners(ctx)
+	case clusterTypeURL:
+		stream, err = m.cdsClient.StreamClusters(ctx)
+	case routeTypeURL:
+		stream, err = m.rdsClient.StreamRoutes(ctx)
+	case endpointTypeURL:
+		stream, err = m.edsClient.StreamEndpoints(ctx)
+	default:
+		return nil, fmt.Errorf("Unsupported request type")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	signal := make(chan *version, 1)
+	signal <- &version{}
+	response := make(chan *Response)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+
+	go send(ctx, &wg, request, response, stream, signal, m.callOptions)
+	go recv(ctx, &wg, response, stream, signal)
+
+	go func() {
+		wg.Wait()
+		closeChannel(signal, response)
+	}()
+
+	return response, nil
+}
+
+func send(
+	ctx context.Context,
+	wg *sync.WaitGroup,
+	request *v2.DiscoveryRequest,
+	response chan *Response,
+	stream grpc.ClientStream,
+	signal chan *version,
+	callOptions CallOptions) {
+	for {
+		select {
+		case sig, ok := <-signal:
+			if !ok {
+				return
+			}
+			request.ResponseNonce = sig.nonce
+			request.VersionInfo = sig.version
+			err := doWithTimeout(func() error {
+				return stream.SendMsg(request)
+			}, callOptions.Timeout)
```

Comment:
shouldn't the context provide the Deadline here?

Reply:
the ctx does not have a deadline at this point. `doWithTimeout ` uses a child context with deadline.

Reply:
I don't see where it uses the child context? the context doesn't seem to be passed into the doWithTimeout method.

Reply:
At your time of comment i was using a timer to achieve timeout. I. have since changed the [code](https://github.com/envoyproxy/xds-relay/pull/46/files#diff-c30e80dfa8b61ea3ccdc1ae6d3716890R226) to use ctx and deadline. The github view here is probably stale.